DataFormats/EgammaCandidates/interface/GsfElectron.h

```diff
@@ -183,11 +183,14 @@ class GsfElectron : public RecoCandidate
     // forward core methods
     SuperClusterRef superCluster() const override { return core()->superCluster() ; }
     GsfTrackRef gsfTrack() const override { return core()->gsfTrack() ; }
-    virtual TrackRef closestTrack() const { return core()->ctfTrack() ; }
     float ctfGsfOverlap() const { return core()->ctfGsfOverlap() ; }
     bool ecalDrivenSeed() const { return core()->ecalDrivenSeed() ; }
     bool trackerDrivenSeed() const { return core()->trackerDrivenSeed() ; }
     virtual SuperClusterRef parentSuperCluster() const { return core()->parentSuperCluster() ; }
+    bool closestTrackValid() const { return closestCtfTrackRef().isAvailable() && closestCtfTrackRef().isNonnull() ; }
+    //methods used for MVA variables
+    float closestTrackChi2() const; 
+    int closestTrackNHits() const; 
```

Comment:
I noticed now that the implementation is ```hitPattern().trackerLayersWithMeasurement()```.
Shouldn't the method be called ```closestTrackNLayers``` ? Otherwise this seems potentially quite confusing.

Reply:
that was what it was named for the MVA variables.  Unfortunately this name is grandfathered and  I dont really want to change it.

Reply:
TrackChi2 is also not a full chi2, but a normalized chi2. Slightly less confusing than the NHits which are actually layers.

Instead of introducing confusingly named methods in general use classes, it seems more practical then to let the code that fills the MVA variables call them to its liking.

Reply:
Please clarify which MVA code had this naming.
Thank you.

I see one case:
https://github.com/cms-sw/cmssw/blob/570d18fd35f56f0e8865674205dacc20b1bd9328/EgammaAnalysis/ElectronTools/src/EGammaMvaEleEstimator.cc#L1155
an internal variable in the code refers to hits, but then when it's passed to the MVA for evaluation, it appropriately refers to layers here
https://github.com/cms-sw/cmssw/blob/570d18fd35f56f0e8865674205dacc20b1bd9328/EgammaAnalysis/ElectronTools/src/EGammaMvaEleEstimator.cc#L256
and here
https://github.com/cms-sw/cmssw/blob/570d18fd35f56f0e8865674205dacc20b1bd9328/EgammaAnalysis/ElectronTools/src/EGammaMvaEleEstimator.cc#L316

Reply:
perhaps adding a ```mva_``` prefix and a clear note that the returned value does not match the name could be acceptable.
Also, if the implementation can be moved inline it will also be more clear.

Reply:
Oh I made my peace with this decided to change it as suggested and do as much work as it takes to get this in. Means I wont do other things of course because my time is finite, but fine. 

In fact, I now named them closestCtfTrack as closestTrack is gone. 

The MVA uses it here: 
https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2.h#L54-L55

Reply:
well that was fun! So turns out including reco::Track header file (necessary to inline the methods as requested) caused AnalysisDataFormats/Egamma not to compile. This is because GsfElectronFwd.h improperly includes GsfElectron.h  (so not really a forward declaration as much as an actual declaration, sigh).  The dictionary of AnalysisDataFormats/EGamma therefore needed the CLHEP dependence in the build file.

And this is a classic example of what looks like small simple change ending up taking significant time.